cli/src/action/login/oauth/google.rs

```diff
@@ -0,0 +1,237 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use oauth2::basic::BasicClient;
+use oauth2::{
+    AccessToken, AuthUrl, AuthorizationCode, ClientId, ClientSecret, CsrfToken, PkceCodeChallenge,
+    RedirectUrl, Scope, TokenResponse, TokenUrl,
+};
+use reqwest::blocking::{self, Client};
+
+use super::callback;
+use super::{OAuth2Provider, OAuth2ProviderError, ProviderType, UserTokens};
+
+pub struct GoogleProvider {
+    client_id: Option<&'static str>,
+    client_secret: Option<&'static str>,
+}
+
+impl OAuth2Provider for GoogleProvider {
+    fn get_tokens(&self) -> Result<UserTokens, OAuth2ProviderError> {
+        let client_id = self.client_id.ok_or_else(|| {
```

Comment:
It seems like we should be able to refactor a lot of this code to be shared between the provider implementations. Would make it easier to add more providers in the future.